### PR TITLE
Truffle config handling

### PR DESCRIPTION
**Pull request by vreturn**
Hey all, I ran into an edge case while configuring a test system (specifically bumping the solidity version used).

The original `truffle.js` looked like this:
```js
module.exports = {
  networks: {
    development: {
      host: '127.0.0.1',
      port: 7545,
      network_id: '*',
    },
  },
}
```

I added solc settings making it look like this:
```js
module.exports = {
  compilers: {
    solc: {
      version: '^0.5.10',
      docker: false,
      settings: {
        optimizer: {
          enabled: false,
        },
      }
    }
  },
  networks: {
    development: {
      host: '127.0.0.1',
      port: 7545,
      network_id: '*',
    },
  },
}
```

The above broke my test system as there is no `coverage` network so my `truffle.js` got replaced by the solidity-coverage default `truffle-config.js`.

By default the entire truffle config is discarded if the `coverage` network is not present. I changed this to instead add the default coverage network to an existing truffle config, if one exists; then write to file. I also restructured how the default config is stored.

This should be backward compatible with those that have defined a `coverage` network, and makes it easier for new users to customize the `truffle.js`.

### CODE REVIEW

File: lib/app.js [hunk 1]
```diff
@@ -102,15 +101,35 @@ class App {
...
+        network_id: '*',
+        port: this.port,
+        gas: gasLimitHex,
+        gasPrice: gasPriceHex
+      };
+
       // Coverage network opts specified: use port if declared
       if (truffleConfig && truffleConfig.networks && truffleConfig.networks.coverage) {
         this.port = truffleConfig.networks.coverage.port || this.port;
         this.network = '--network coverage';
```
Comment (cgewecke) [inline lib/app.js]: The `this.network` assignment should come out of the `if` block here so it's always set. Maybe just right before this piece of logic. It gets appended to the `truffle test` command later and I think the unit-tests are failing because they're not running `truffle test --network coverage` as they should.

File: lib/app.js [hunk 2]
```diff
@@ -102,15 +101,35 @@ class App {
...
 
       // No coverage network defaults to the dev network on port 8555, high gas / low price.
+      } else if (truffleConfig) {
```
Comment (cgewecke) [inline lib/app.js]: Here, can we make this just plain `else`? Truffle always needs to find a config anyway so it's probably there. To be safe we could initialize truffleConfig to `{}` above here where it's declared.

File: lib/app.js [hunk 3]
```diff
@@ -102,15 +101,35 @@ class App {
...
+        // Instead of writing the default, put the coverage in the existing
+        // config
+        if (!truffleConfig.networks) truffleConfig.networks = {};
+        truffleConfig.networks.coverage = coverageNetwork;
+        const configString = `module.exports = ${JSON.stringify(truffleConfig)}`;
+        fs.writeFileSync(`${this.coverageDir}/truffle-config.js`, configString);
       } else {
```
Comment (cgewecke) [inline lib/app.js]: And then this whole `else` block can be deleted I think...

File: lib/app.js (hunk 2 above)
Comment (cgewecke) [reply lib/app.js]: And if truffleConfig is initialized by default the condition check above this else can be simplified to:
```javascript
if (truffleConfig.networks && truffleConfig.networks.coverage)
```

File: lib/app.js [hunk 4]
```diff
@@ -102,15 +101,35 @@ class App {
       if      (shell.test('-e', oldTrufflePath)) truffleConfig = reqCwd.silent(oldTrufflePath);
       else if (shell.test('-e', newTrufflePath)) truffleConfig = reqCwd.silent(newTrufflePath);
```
Comment (cgewecke) [inline lib/app.js]: @vreturn one last thing - believe we could also resolve #359 in this PR if we reverse the order of the config type tests here - e.g. test for newTrufflePath first because that's what Truffle prefers when both configs are present.

**Comment by vreturn** [comment]
Hey @cgewecke thanks for the notes.

Refactored and fixed the e2e tests. The new scheme for remote git url generation _should_ work for all other pr's, but if you run into bugs I'd check the diffs here.
